tachyon/math/matrix/matrix_utils.h

```diff
@@ -74,6 +77,49 @@ std::vector<PackedPrimeField> PackRowVertically(
   });
 }
 
+// Expands a |RowMajorMatrix|'s rows from |rows| to |rows|^{|added_bits|},
+// moving values from row |i| to row |i|^{|added_bits|}. All new entries are set
+// to F::Zero()
+template <typename F>
+void ExpandWithZeroPad(RowMajorMatrix<F>& mat, size_t added_bits) {
+  if (added_bits == 0) {
+    return;
+  }
+
+  size_t original_rows = static_cast<size_t>(mat.rows());
+  size_t new_rows = static_cast<size_t>(mat.rows()) << added_bits;
+  size_t cols = static_cast<size_t>(mat.cols());
+
+  RowMajorMatrix<F> padded =
+      RowMajorMatrix<F>::Constant(new_rows, cols, F::Zero());
+
+  OPENMP_PARALLEL_FOR(size_t row = 0; row < original_rows; ++row) {
+    size_t padded_row_index = row << added_bits;
+    // TODO(ashjeong): Check if moved properly
+    padded.row(padded_row_index) = std::move(mat.row(row));
+  }
+  mat = std::move(padded);
+}
+
+// Swaps rows of a |RowMajorMatrix| such that each row is changed to the row
+// accessed with the reversed bits of the current index
+template <typename F>
+void ReverseMatrixIndexBits(RowMajorMatrix<F>& mat) {
+  size_t rows = static_cast<size_t>(mat.rows());
```

Comment:
I've kept this the same as before, since it would have to be casted multiple times for "SafeLog2Ceiling" and "BitRev," so holding it as a "size_t" appears to be more efficient.